### PR TITLE
new version 0.9.7 with mode test support

### PR DESCRIPTION
**Pull request by splanquart**

### CODE REVIEW

File: ipn.php [hunk 1]
```diff
@@ -100,31 +124,48 @@ function getallheaders()
 
 	if ($data && $bool_sign)
 	{
-		$cart = new Cart($data['custom_data']);
+		/** Data is an object */
+		$cart = new Cart($data->custom_data);
+		$address = new Address((int)$cart->id_address_invoice);
```
Comment (Quetzacoalt91) [inline ipn.php]: You should check if your cart (and address) are properly loaded with ```Validate::isLoadedObject()```

File: payplug.php [hunk 2]
```diff
@@ -24,43 +24,47 @@
...
+		// Update version
+		$this->version = '0.9.7';
 		$this->author = 'PayPlug';
 		$this->module_key = '1ee28a8fb5e555e274bd8c2e1c45e31a';
+
 		parent::__construct();
 
 		// Backward compatibility
-		require(_PS_MODULE_DIR_.$this->name.'/backward_compatibility/backward.php');
+		if (version_compare(_PS_VERSION_, '1.4', '>'))
```
Comment (Quetzacoalt91) [inline payplug.php]: You need only the backward compatibility on PS 1.4, so a better condition would be ```if (version_compare(_PS_VERSION_, '1.5', '<'))```

File: payplug.php [hunk 3]
```diff
@@ -100,22 +103,102 @@ private function upgrade()
...
 	public function install()
 	{
 		if (version_compare(_PS_VERSION_, '1.4', '<') || !parent::install() || !$this->registerHook('payment') || !$this->registerHook('paymentReturn'))
 			return false;
+
+		// add hook for 1.6
+		if (version_compare(_PS_VERSION_, '1.5', '<'))
+		{
+			if (!$this->registerHook('header'))
+				return false;
```
Comment (Quetzacoalt91) [inline payplug.php]:
```php
	if (!$this->registerHook('header'))
		return false;
```
will work on PS 1.4, but 1.5 and 1.6 too. You do not need to check the version for a displayHeader case.

File: payplug.php [hunk 4]
```diff
@@ -263,4 +445,29 @@ public function hookPaymentReturn()
...
+	 */
+	public function hookHeader()
+	{
+		return $this->hookDisplayHeader();
+	}
+
+	/**
+	 * Hook for >= 1.5
+	 */
+	public function hookDisplayHeader()
```
Comment (Quetzacoalt91) [inline payplug.php]: Same comment about the hooks.

You do not need ```public function hookDisplayHeader()``` if ```public function hookHeader()``` exists.

File: installPayplug.php [hunk 5]
```diff
@@ -62,84 +93,116 @@ public function deleteConfig()
...
-				if ($order_state->add())
+				// Update OS PayPlug payment
+				if (defined('_'.$cfg.'_') || Payplug::getConfiguration($cfg))
 				{
-					$source = dirname(__FILE__).'/logo.gif';
-					$destination = dirname(__FILE__).'/../../img/os/'.(int)$order_state->id.'.gif';
-					copy($source, $destination);
+					// If is in configuration (since 1.5)
+					if (!($os = Configuration::get($cfg)))
+						$os = defined('_'.$cfg.'_');
```
Comment (Quetzacoalt91) [inline installPayplug.php]: ```defined(...)``` with return a boolean. Is it expected ?

File: payplug.php [hunk 6]
```diff
@@ -100,22 +103,102 @@ private function upgrade()
 					// If is defined
 					else
 						$os_payment = defined('_PS_OS_PAYMENT_');
-					Configuration::updateValue('PAYPLUG_ORDER_STATE_PAID', (int)$os_payment);
+
+					Payplug::updateConfiguration('PAYPLUG_ORDER_STATE_PAID', (int)$os_payment);
 				}
 			}
+			// Add test status && add hook
+			if ($version === false || version_compare($version, '0.9.7', '<'))
```
Comment (Quetzacoalt91) [inline payplug.php]: Do you know that PrestaShop can already auto-upgrade the modules ?
http://doc.prestashop.com/display/PS15/Auto-updating+modules

**Comment by GuigZ-** [comment]
See https://github.com/PrestaShop/payplug/pull/5 for the respone !

**Comment by 202-ecommerce** [comment]
See https://github.com/PrestaShop/payplug/pull/5 for the respone !

**Pull request merged**